msal/src/main/java/com/microsoft/identity/client/PublicClientApplication.java

```diff
@@ -1618,6 +1623,49 @@ public void onError(MsalException exception) {
         }
     }
 
+    public void deviceCodeFlow(@Nullable String[] scopes, @NonNull final DeviceCodeFlowCallback callback) {
+        // Create a DeviceCodeFlowCommandParameters object that takes in the desired scopes and the callback object
+        // Use CommandParametersAdapter
+        final DeviceCodeFlowCommandParameters commandParameters = CommandParametersAdapter
+                    .createDeviceCodeFlowCommandParameters(
+                            mPublicClientConfiguration,
+                            mPublicClientConfiguration.getOAuth2TokenCache(),
+                            scopes);
+
+        // Create a CommandCallback object from the DeviceCodeFlowCallback object
+        final DeviceCodeFlowCommandCallback dcfCallback = getDeviceCodeFlowCommandCallback(callback);
+
+        // Attempt protocol
+        try {
+            // Create a DeviceCodeFlowCommand object
+            // Pass the command parameters, default controller, and command callback
+            // Telemetry with DEVICE_CODE_FLOW_CALLBACK
+            final DeviceCodeFlowCommand deviceCodeFlowCommand = new DeviceCodeFlowCommand(
+                    commandParameters,
+                    MSALControllerFactory.getDefaultController(
```

Comment:
Looking at your design doc, I see you called out that (at least for now), Device Code Flow is only supported for Local Msal and NOT supported via the Broker. Over here I see the `DeviceCodeFlowCommand` consuming the Default Controller by calling `getDefaultController` - that's contradicting what's in the design doc. `getDefaultController` returns the `BrokerMsalController` is a broker is installed on the device.

What is our plan in this case? Do we let flow go over to the broker and then throw some exception indicating that the DCF is not supported? Or wouldn't it be better for the `DeviceCodeFlowCommand` to just consume the `LocalMsalController` (even if a broker is present on the device) to always use the local MSAL flow for the `DeviceCodeFlowCommand` ?

Reply:
Yes, Scope is limited to MSAL. At the start of the design process, I was going to look into how to limit the command to a single controller instead of using default controller, but I forgot to do so as I was creating the design document and focused on other things. But yes, ideally I would want the API to just use LocalMsalController even if Broker is installed on the device (at least until Broker support is implemented).